File: aoc15/ocaml/day_03/day_03.ml
type pos = { x : int; y : int }

module Grid = Map.Make (struct
  type t = pos

  let compare = compare
end)
;;

let input = Input.chars "input" in

let insert_or_increase v =
  match v with Some x -> Some (x + 1) | None -> Some 1
in
let move (grid, pos) dir =
  let next =
    match dir with
    | '^' -> { x = pos.x; y = pos.y + 1 }
    | 'v' -> { x = pos.x; y = pos.y - 1 }
    | '>' -> { x = pos.x + 1; y = pos.y }
    | '<' -> { x = pos.x - 1; y = pos.y }
    | _ -> pos
  in
  (Grid.update next insert_or_increase grid, next)
in
let visited dirs =
  let houses = Grid.empty in
  let pos = { x = 0; y = 0 } in
  let houses = Grid.add pos 1 houses in
  let visits, _ = List.fold_left move (houses, pos) dirs in
  visits
in

let first = Grid.cardinal (visited input) in
Printf.printf "First: %d\n" first;

let split lst = List.fold_right (fun x (l, r) -> (x :: r, l)) lst ([], []) in
let santa, robot = split input in
let merge _ val1 val2 =
  match (val1, val2) with
  | Some x, Some y -> Some (x + y)
  | None, x | x, None -> x
in
let final_grid = Grid.merge merge (visited santa) (visited robot) in

let second = Grid.cardinal final_grid in
Printf.printf "Second: %d\n" second

File: aoc15/go/day_03/day_03.go
package main

import (
	"fmt"
	"os"
)

type House struct {
	x, y int
}

func deliver(input []byte, Houses map[House]bool, start, step int) {
	pos := House{0, 0}
	Houses[pos] = true
	for i := start; i < len(input); i += step {
		switch input[i] {
		case '^':
			pos = House{pos.x, pos.y + 1}
		case 'v':
			pos = House{pos.x, pos.y - 1}
		case '>':
			pos = House{pos.x + 1, pos.y}
		case '<':
			pos = House{pos.x - 1, pos.y}
		}
		Houses[pos] = true
	}
}

func main() {
	input, _ := os.ReadFile("input")

	first := make(map[House]bool)
	deliver(input, first, 0, 1)
	second := make(map[House]bool)
	deliver(input, second, 0, 2)
	deliver(input, second, 1, 2)
	fmt.Println("First:", len(first))
	fmt.Println("Second:", len(second))
}
